src/solver.rs

```diff
@@ -171,6 +173,58 @@ pub fn resolve<P: Package, V: Version>(
     }
 }
 
+/// An enum used by [DependencyProvider] that holds information about package dependencies.
+/// For each [Package] there is a [Range] of concrete versions it allows as a dependency.
+#[derive(Clone)]
+pub enum Dependencies<P: Package, V: Version> {
+    /// Package dependencies are unavailable.
+    Unavailable,
+    /// Container for all available package versions.
+    Known(AllowedVersions<P, V>),
+}
+
+/// Subtype of [Dependencies] which holds information about
+/// all possible versions a given package can accept.
+/// There is a difference in semantics between an empty [Map<P, Range<V>>](crate::type_aliases::Map)
+/// inside [AllowedVersions] and [Dependencies::Unavailable]:
+/// the former means the package has no dependencies and it is a known fact,
+/// while the latter means they could not be fetched by [DependencyProvider].
+#[derive(Debug, Clone)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(transparent)
+)]
+pub struct AllowedVersions<P: Package, V: Version>(Map<P, Range<V>>);
```

Comment:
I don't like much that type. It feels too much

Reply:
Or maybe that's the type that should be named `Dependencies` and the other should be renamed since those are not dependencies but the result of a dependency request

Reply:
I'll sleep on the naming.